File: api/routes/admin-subscriptions.js
const express = require('express');
const { spawn } = require('child_process');
const path = require('path');
const { authenticateToken, requireRole } = require('../middleware/auth');

const router = express.Router();

// All routes require authentication
router.use(authenticateToken);

/**
 * GET /api/admin/subscriptions
 * List all subscriptions
 */
router.get('/', async (req, res) => {
  try {
    const python = spawn('python', ['-c', `
import sys
import json
sys.path.insert(0, '${path.join(__dirname, '..', '..')}')
from src.subscription_manager import SubscriptionManager
manager = SubscriptionManager()
subscriptions = manager.list_subscriptions()
print(json.dumps({'subscriptions': subscriptions}))
    `]);

    let stdout = '';
    let stderr = '';
    python.stdout.on('data', (data) => stdout += data);
    python.stderr.on('data', (data) => stderr += data);
    python.on('close', (code) => {
      if (code === 0) {
        try {
          const result = JSON.parse(stdout);
          res.json(result);
        } catch (err) {
          console.error('Failed to parse subscriptions output:', stdout);
          res.status(500).json({ error: 'Failed to parse subscriptions data' });
        }
      } else {
        console.error('Failed to list subscriptions:', stderr);
        res.status(500).json({ error: 'Failed to fetch subscriptions' });
      }
    });
  } catch (error) {
    console.error('Error fetching subscriptions:', error);
    res.status(500).json({ error: 'Failed to fetch subscriptions' });
  }
});

/**
 * POST /api/admin/subscriptions
 * Add a new subscription
 * Requires: admin role or higher
 */
router.post('/', requireRole(['admin', 'super_admin']), async (req, res) => {
  try {
    const { source_url, source_name, priority, check_interval_hours } = req.body;

    if (!source_url) {
      return res.status(400).json({ error: 'source_url is required' });
    }

    const python = spawn('python', ['-c', `
import sys
import json
sys.path.insert(0, '${path.join(__dirname, '..', '..')}')
from src.subscription_manager import SubscriptionManager
manager = SubscriptionManager()
result = manager.add_subscription(
    source_url=${JSON.stringify(source_url)},
    source_name=${source_name ? JSON.stringify(source_name) : 'None'},
    priority=${priority != null ? JSON.stringify(priority) : 'None'},
    check_interval_hours=${check_interval_hours != null ? JSON.stringify(check_interval_hours) : 'None'}
)
print(json.dumps({'subscription': result}))
    `]);

    let stdout = '';
    let stderr = '';
    python.stdout.on('data', (data) => stdout += data);
    python.stderr.on('data', (data) => stderr += data);
    python.on('close', (code) => {
      if (code === 0) {
        try {
          const result = JSON.parse(stdout);
          res.status(201).json(result);
        } catch (err) {
          console.error('Failed to parse add subscription output:', stdout);
          res.status(500).json({ error: 'Failed to parse subscription data' });
        }
      } else {
        console.error('Failed to add subscription:', stderr);
        res.status(500).json({ error: 'Failed to add subscription' });
      }
    });
  } catch (error) {
    console.error('Error adding subscription:', error);
    res.status(500).json({ error: 'Failed to add subscription' });
  }
});

/**
 * GET /api/admin/subscriptions/:id
 * Get subscription details
 */
router.get('/:id', async (req, res) => {
  try {
    const python = spawn('python', ['-c', `
import sys
import json
sys.path.insert(0, '${path.join(__dirname, '..', '..')}')
from src.subscription_manager import SubscriptionManager
manager = SubscriptionManager()
subscription = manager.get_subscription('${req.params.id}')
print(json.dumps({'subscription': subscription}))
    `]);

    let stdout = '';
    let stderr = '';
    python.stdout.on('data', (data) => stdout += data);
    python.stderr.on('data', (data) => stderr += data);
    python.on('close', (code) => {
      if (code === 0) {
        try {
          const result = JSON.parse(stdout);
          if (!result.subscription) {
            return res.status(404).json({ error: 'Subscription not found' });
          }
          res.json(result);
        } catch (err) {
          console.error('Failed to parse subscription output:', stdout);
          res.status(500).json({ error: 'Failed to parse subscription data' });
        }
      } else {
        console.error('Failed to fetch subscription:', stderr);
        res.status(500).json({ error: 'Failed to fetch subscription' });
      }
    });
  } catch (error) {
    console.error('Error fetching subscription:', error);
    res.status(500).json({ error: 'Failed to fetch subscription details' });
  }
});

/**
 * PUT /api/admin/subscriptions/:id
 * Update a subscription
 * Requires: admin role or higher
 */
router.put('/:id', requireRole(['admin', 'super_admin']), async (req, res) => {
  try {
    const { priority, check_interval_hours, source_name } = req.body;

    const python = spawn('python', ['-c', `
import sys
import json
sys.path.insert(0, '${path.join(__dirname, '..', '..')}')
from src.subscription_manager import SubscriptionManager
manager = SubscriptionManager()
updates = {}
${priority != null ? `updates['priority'] = ${JSON.stringify(priority)}` : ''}
${check_interval_hours != null ? `updates['check_interval_hours'] = ${JSON.stringify(check_interval_hours)}` : ''}
${source_name ? `updates['source_name'] = ${JSON.stringify(source_name)}` : ''}
result = manager.update_subscription('${req.params.id}', **updates)
print(json.dumps({'subscription': result}))
    `]);

    let stdout = '';
    let stderr = '';
    python.stdout.on('data', (data) => stdout += data);
    python.stderr.on('data', (data) => stderr += data);
    python.on('close', (code) => {
      if (code === 0) {
        try {
          const result = JSON.parse(stdout);
          if (!result.subscription) {
            return res.status(404).json({ error: 'Subscription not found' });
          }
          res.json(result);
        } catch (err) {
          console.error('Failed to parse update subscription output:', stdout);
          res.status(500).json({ error: 'Failed to parse subscription data' });
        }
      } else {
        console.error('Failed to update subscription:', stderr);
        res.status(500).json({ error: 'Failed to update subscription' });
      }
    });
  } catch (error) {
    console.error('Error updating subscription:', error);
    res.status(500).json({ error: 'Failed to update subscription' });
  }
});

/**
 * DELETE /api/admin/subscriptions/:id
 * Delete a subscription
 * Requires: super_admin role
 */
router.delete('/:id', requireRole(['super_admin']), async (req, res) => {
  try {
    const python = spawn('python', ['-c', `
import sys
import json
sys.path.insert(0, '${path.join(__dirname, '..', '..')}')
from src.subscription_manager import SubscriptionManager
manager = SubscriptionManager()
result = manager.delete_subscription('${req.params.id}')
print(json.dumps({'success': result}))
    `]);

    let stdout = '';
    let stderr = '';
    python.stdout.on('data', (data) => stdout += data);
    python.stderr.on('data', (data) => stderr += data);
    python.on('close', (code) => {
      if (code === 0) {
        try {
          const result = JSON.parse(stdout);
          if (!result.success) {
            return res.status(404).json({ error: 'Subscription not found' });
          }
          res.json({ message: 'Subscription deleted successfully' });
        } catch (err) {
          console.error('Failed to parse delete subscription output:', stdout);
          res.status(500).json({ error: 'Failed to parse response' });
        }
      } else {
        console.error('Failed to delete subscription:', stderr);
        res.status(500).json({ error: 'Failed to delete subscription' });
      }
    });
  } catch (error) {
    console.error('Error deleting subscription:', error);
    res.status(500).json({ error: 'Failed to delete subscription' });
  }
});

/**
 * POST /api/admin/subscriptions/:id/pause
 * Pause a subscription
 * Requires: admin role or higher
 */
router.post('/:id/pause', requireRole(['admin', 'super_admin']), async (req, res) => {
  try {
    const python = spawn('python', ['-c', `
import sys
import json
sys.path.insert(0, '${path.join(__dirname, '..', '..')}')
from src.subscription_manager import SubscriptionManager
manager = SubscriptionManager()
result = manager.pause_subscription('${req.params.id}')
print(json.dumps({'subscription': result}))
    `]);

    let stdout = '';
    let stderr = '';
    python.stdout.on('data', (data) => stdout += data);
    python.stderr.on('data', (data) => stderr += data);
    python.on('close', (code) => {
      if (code === 0) {
        try {
          const result = JSON.parse(stdout);
          if (!result.subscription) {
            return res.status(404).json({ error: 'Subscription not found' });
          }
          res.json({ message: 'Subscription paused successfully', ...result });
        } catch (err) {
          console.error('Failed to parse pause subscription output:', stdout);
          res.status(500).json({ error: 'Failed to parse response' });
        }
      } else {
        console.error('Failed to pause subscription:', stderr);
        res.status(500).json({ error: 'Failed to pause subscription' });
      }
    });
  } catch (error) {
    console.error('Error pausing subscription:', error);
    res.status(500).json({ error: 'Failed to pause subscription' });
  }
});

/**
 * POST /api/admin/subscriptions/:id/resume
 * Resume a paused subscription
 * Requires: admin role or higher
 */
router.post('/:id/resume', requireRole(['admin', 'super_admin']), async (req, res) => {
  try {
    const python = spawn('python', ['-c', `
import sys
import json
sys.path.insert(0, '${path.join(__dirname, '..', '..')}')
from src.subscription_manager import SubscriptionManager
manager = SubscriptionManager()
result = manager.resume_subscription('${req.params.id}')
print(json.dumps({'subscription': result}))
    `]);

    let stdout = '';
    let stderr = '';
    python.stdout.on('data', (data) => stdout += data);
    python.stderr.on('data', (data) => stderr += data);
    python.on('close', (code) => {
      if (code === 0) {
        try {
          const result = JSON.parse(stdout);
          if (!result.subscription) {
            return res.status(404).json({ error: 'Subscription not found' });
          }
          res.json({ message: 'Subscription resumed successfully', ...result });
        } catch (err) {
          console.error('Failed to parse resume subscription output:', stdout);
          res.status(500).json({ error: 'Failed to parse response' });
        }
      } else {
        console.error('Failed to resume subscription:', stderr);
        res.status(500).json({ error: 'Failed to resume subscription' });
      }
    });
  } catch (error) {
    console.error('Error resuming subscription:', error);
    res.status(500).json({ error: 'Failed to resume subscription' });
  }
});

/**
 * POST /api/admin/subscriptions/:id/check
 * Trigger immediate poll for a subscription
 * Requires: admin role or higher
 */
router.post('/:id/check', requireRole(['admin', 'super_admin']), async (req, res) => {
  try {
    const python = spawn('python', ['-c', `
import sys
import json
sys.path.insert(0, '${path.join(__dirname, '..', '..')}')
from src.pipeline_scheduler import PipelineScheduler
from src.database import Database
scheduler = PipelineScheduler()
sub = scheduler.sub_manager.get_subscription('${req.params.id}')
if sub:
    videos = scheduler._fetch_channel_videos(sub)
    print(json.dumps({
        'success': True,
        'subscription_id': '${req.params.id}',
        'videos_found': len(videos) if videos else 0,
        'videos': videos[:10] if videos else []
    }))
else:
    print(json.dumps({'success': False, 'error': 'Subscription not found'}))
    `]);

    let stdout = '';
    let stderr = '';
    python.stdout.on('data', (data) => stdout += data);
    python.stderr.on('data', (data) => stderr += data);
    python.on('close', (code) => {
      if (code === 0) {
        try {
          const result = JSON.parse(stdout);
          if (!result.success) {
            return res.status(404).json({ error: result.error || 'Subscription not found' });
          }
          res.json(result);
        } catch (err) {
          console.error('Failed to parse check subscription output:', stdout);
          res.status(500).json({ error: 'Failed to parse response' });
        }
      } else {
        console.error('Failed to check subscription:', stderr);
        res.status(500).json({ error: 'Failed to trigger subscription check' });
      }
    });
  } catch (error) {
    console.error('Error checking subscription:', error);
    res.status(500).json({ error: 'Failed to trigger subscription check' });
  }
});

/**
 * POST /api/admin/subscriptions/:id/refresh
 * Refresh a subscription: fetch latest videos, queue new ones, skip old ones.
 * Fetches the N most recent videos, skips already-processed ones, queues the rest.
 * Requires: admin role or higher
 */
router.post('/:id/refresh', requireRole(['admin', 'super_admin']), async (req, res) => {
  try {
    const subscriptionId = req.params.id;
    const python = spawn('python', ['-c', `
import sys
import json
sys.path.insert(0, '${path.join(__dirname, '..', '..')}')
sys.path.insert(0, '${path.join(__dirname, '..', '..', 'src')}')
from pipeline_scheduler import PipelineScheduler
scheduler = PipelineScheduler()
try:
    result = scheduler.refresh_subscription('${subscriptionId}')
    print(json.dumps({'success': True, **result}))
except ValueError as e:
    print(json.dumps({'success': False, 'error': str(e)}))
except Exception as e:
    print(json.dumps({'success': False, 'error': str(e)}))
    `]);

    let stdout = '';
    let stderr = '';
    python.stdout.on('data', (data) => stdout += data);
    python.stderr.on('data', (data) => stderr += data);
    python.on('close', (code) => {
      if (code === 0) {
        try {
          const result = JSON.parse(stdout);
          if (!result.success) {
            const status = result.error && result.error.includes('not found') ? 404 : 500;
            return res.status(status).json({ error: result.error });
          }
          res.json(result);
        } catch (err) {
          console.error('Failed to parse refresh subscription output:', stdout);
          res.status(500).json({ error: 'Failed to parse response' });
        }
      } else {
        console.error('Failed to refresh subscription:', stderr);
        res.status(500).json({ error: 'Failed to refresh subscription' });
      }
    });
  } catch (error) {
    console.error('Error refreshing subscription:', error);
    res.status(500).json({ error: 'Failed to refresh subscription' });
  }
});

module.exports = router;

File: api/routers/admin_subscriptions.py
"""Admin subscription management endpoints for FastAPI."""

import sys
from pathlib import Path
from typing import Optional

from fastapi import APIRouter, HTTPException, Query, Depends
from pydantic import BaseModel

from .admin import get_current_user, require_role

# Add src to path for imports
possible_src_paths = [
    Path("/app/src"),
    Path(__file__).parent.parent.parent / "src",
    Path(__file__).parent.parent / "src",
]
for src_path in possible_src_paths:
    if src_path.exists() and str(src_path.resolve()) not in sys.path:
        sys.path.insert(0, str(src_path.resolve()))
        break

router = APIRouter(prefix="/api/admin/subscriptions", tags=["Admin - Subscriptions"])


# Request/response models
class AddSubscriptionRequest(BaseModel):
    source_url: str
    source_name: Optional[str] = None
    priority: Optional[int] = 5
    check_interval_hours: Optional[int] = 12


class UpdateSubscriptionRequest(BaseModel):
    source_name: Optional[str] = None
    priority: Optional[int] = None
    check_interval_hours: Optional[int] = None


def _get_subscription_manager():
    """Get SubscriptionManager instance."""
    from database import get_database
    from subscription_manager import SubscriptionManager
    db = get_database()
    return SubscriptionManager(db)


@router.get(
    "",
    summary="List subscriptions",
    description="List all YouTube channel/playlist subscriptions.",
)
async def list_subscriptions(
    active_only: bool = Query(True, description="Only return active subscriptions"),
    user: dict = Depends(get_current_user),
):
    """List all subscriptions."""
    manager = _get_subscription_manager()
    subscriptions = manager.list_subscriptions(active_only=active_only)
    return {"subscriptions": subscriptions}


@router.post(
    "",
    status_code=201,
    summary="Add subscription",
    description="Add a new YouTube channel or playlist subscription. Requires admin role.",
)
async def add_subscription(
    request: AddSubscriptionRequest,
    user: dict = Depends(require_role(["admin", "super_admin"])),
):
    """Add a new subscription."""
    manager = _get_subscription_manager()
    try:
        subscription = manager.add_subscription(
            source_url=request.source_url,
            source_name=request.source_name,
            priority=request.priority,
            check_interval_hours=request.check_interval_hours,
        )
        return {"subscription": subscription}
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.get(
    "/{subscription_id}",
    summary="Get subscription",
    description="Get details of a specific subscription.",
)
async def get_subscription(
    subscription_id: str,
    user: dict = Depends(get_current_user),
):
    """Get subscription by ID."""
    manager = _get_subscription_manager()
    subscription = manager.get_subscription(subscription_id)
    if not subscription:
        raise HTTPException(status_code=404, detail="Subscription not found")
    return {"subscription": subscription}


@router.put(
    "/{subscription_id}",
    summary="Update subscription",
    description="Update subscription settings. Requires admin role.",
)
async def update_subscription(
    subscription_id: str,
    request: UpdateSubscriptionRequest,
    user: dict = Depends(require_role(["admin", "super_admin"])),
):
    """Update a subscription."""
    manager = _get_subscription_manager()

    # Build updates dict from non-None fields
    updates = {}
    if request.source_name is not None:
        updates["source_name"] = request.source_name
    if request.priority is not None:
        updates["priority"] = request.priority
    if request.check_interval_hours is not None:
        updates["check_interval_hours"] = request.check_interval_hours

    if not updates:
        raise HTTPException(status_code=400, detail="No fields to update")

    success = manager.update_subscription(subscription_id, **updates)
    if not success:
        raise HTTPException(status_code=404, detail="Subscription not found")

    subscription = manager.get_subscription(subscription_id)
    return {"subscription": subscription}


@router.delete(
    "/{subscription_id}",
    summary="Delete subscription",
    description="Delete a subscription. Requires super_admin role.",
)
async def delete_subscription(
    subscription_id: str,
    user: dict = Depends(require_role(["super_admin"])),
):
    """Delete a subscription."""
    manager = _get_subscription_manager()
    success = manager.delete_subscription(subscription_id)
    if not success:
        raise HTTPException(status_code=404, detail="Subscription not found")
    return {"message": "Subscription deleted successfully"}


@router.post(
    "/{subscription_id}/pause",
    summary="Pause subscription",
    description="Pause a subscription. Requires admin role.",
)
async def pause_subscription(
    subscription_id: str,
    user: dict = Depends(require_role(["admin", "super_admin"])),
):
    """Pause a subscription."""
    manager = _get_subscription_manager()
    success = manager.pause_subscription(subscription_id)
    if not success:
        raise HTTPException(status_code=404, detail="Subscription not found")
    subscription = manager.get_subscription(subscription_id)
    return {"message": "Subscription paused successfully", "subscription": subscription}


@router.post(
    "/{subscription_id}/resume",
    summary="Resume subscription",
    description="Resume a paused subscription. Requires admin role.",
)
async def resume_subscription(
    subscription_id: str,
    user: dict = Depends(require_role(["admin", "super_admin"])),
):
    """Resume a paused subscription."""
    manager = _get_subscription_manager()
    success = manager.resume_subscription(subscription_id)
    if not success:
        raise HTTPException(status_code=404, detail="Subscription not found")
    subscription = manager.get_subscription(subscription_id)
    return {"message": "Subscription resumed successfully", "subscription": subscription}


@router.post(
    "/{subscription_id}/check",
    summary="Check subscription now",
    description="Trigger immediate poll for new videos. Requires admin role.",
)
async def check_subscription(
    subscription_id: str,
    user: dict = Depends(require_role(["admin", "super_admin"])),
):
    """Trigger immediate check for new videos."""
    manager = _get_subscription_manager()
    subscription = manager.get_subscription(subscription_id)
    if not subscription:
        raise HTTPException(status_code=404, detail="Subscription not found")

    try:
        from pipeline_scheduler import PipelineScheduler
        scheduler = PipelineScheduler()
        videos = scheduler._fetch_channel_videos(subscription)
        return {
            "success": True,
            "subscription_id": subscription_id,
            "videos_found": len(videos) if videos else 0,
            "videos": videos[:10] if videos else [],
        }
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Failed to check subscription: {str(e)}"
        )


@router.post(
    "/{subscription_id}/refresh",
    summary="Refresh subscription",
    description=(
        "Fetch the latest videos from YouTube, queue the most recent ones "
        "for analysis, skip already-processed videos, and mark older videos "
        "as skipped for admin visibility. Requires admin role."
    ),
)
async def refresh_subscription(
    subscription_id: str,
    user: dict = Depends(require_role(["admin", "super_admin"])),
):
    """Refresh a subscription: fetch latest, queue new, skip old."""
    manager = _get_subscription_manager()
    subscription = manager.get_subscription(subscription_id)
    if not subscription:
        raise HTTPException(status_code=404, detail="Subscription not found")

    try:
        from pipeline_scheduler import PipelineScheduler
        from database import get_database
        db = get_database()
        scheduler = PipelineScheduler(db=db)
        result = scheduler.refresh_subscription(subscription_id)
        return {"success": True, **result}
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Failed to refresh subscription: {str(e)}"
        )
